nestedtensor/nested/masking.py

```diff
@@ -1,111 +1,184 @@
 import torch
 import torch.nn.functional as F
 import numbers
-
 import collections
 
 from . import creation
+from . import utils
+import nestedtensor
 
 TensorMask = collections.namedtuple('TensorMask', 'tensor mask')
 
-# TODO: Can use masked_select?
-
 def nested_tensor_from_padded_tensor(tensor, nested_dim=None, padding=-1):
     mask = (tensor != padding)
     return nested_tensor_from_tensor_mask(tensor, mask, nested_dim)
 
-"""
-Given a tensor t of size N_1 x N_2 x ... N_n and mask m of size M_1 x M_2 x ... M_d where d <= n
-return a NestedTensor of dimensionality d where subtensor t[i_1][i_2]...[i_d] is included as a Tensor constituent if m[i_1][i_2]...[i_d].
-The resulting NestedTensor will be of dimension n.
-Unless specific, it is attempted to yield a NestedTensor of minimal nested dimension.
-"""
-def nested_tensor_from_tensor_mask(tensor, mask, nested_dim=None):
-    # TODO: Should be possible with views only.
-    if not mask.dim() > 0:
-        raise RuntimeError("Mask has to have dimention > 0")
-
-    # TODO: Need to define empty-list and 0 numel semantics
-    # TODO: Test edge-cases and kwargs
-    def _merge(tensors, nested_dim=None):
-        assert len(tensors) > 0
-
-        def _mergable_tensors(tensors, nested_dim):
-            if nested_dim is None:
-                size0 = tensors[0].size()
-                compliant = all(torch.is_tensor(t) for t in tensors)
-                compliant = all(size0 == t.size()
-                                for t in tensors) and compliant
-                return compliant
-            else:
-                return nested_dim < 1
 
-        if _mergable_tensors(tensors, nested_dim):
-            return torch.stack(tensors)
-        else:
-            return creation.nested_tensor(tensors)
+# Constructs nested tensor from passed tensor and mask.
+def nested_tensor_from_tensor_mask(tensor, mask, nested_dim=1):
+    if tensor is None:
+        raise RuntimeError("Tensor can't be undefined (None).")
+
+    if mask is None:
+        raise RuntimeError("Mask can't be undefined (None).")
+
+    # Scalar was passed
+    if tensor.dim() == 0:
+        raise RuntimeError("Can't construct nested tensor from a scalar.")
+
+    if nested_dim == 0:
+        raise RuntimeError("Nested dimension can't be 0.")
 
-    if nested_dim is not None and mask.dim() < nested_dim:
-        raise ValueError("Mask dimension ({0}) is too small to construct nested tensor with nested dimension of {1}".format(mask.dim(), nested_dim))
+    if nested_dim is not None and nested_dim > tensor.dim():
+        raise RuntimeError("Nested dimension ({0}) can't be bigger than data tensor dimension ({1}).".format(nested_dim, tensor.dim()))
 
-    if mask.dim() == 1:
-        tensors = [tensor[i] if mask[i]
-                   else None for i in range(len(mask))]
-        tensors = list(filter(lambda x: x is not None, tensors))
+    if tensor.numel() == 0 and mask.numel() != 0:
```

Comment:
You could check that ```tensor.size()[:mask.dim()] == mask.size()```

Reply:
this code doesnt do the same thing + i find mine to be more readable